flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java

```diff
@@ -0,0 +1,143 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.changelog.StateChangelogStorage;
+import org.apache.flink.runtime.state.changelog.StateChangelogStorageLoader;
+import org.apache.flink.util.ShutdownHookUtil;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class holds the all {@link StateChangelogStorage} objects for a task executor (manager). No
+ * thread-safe.
+ */
+public class TaskExecutorStateChangelogStoragesManager {
+
+    /** Logger for this class. */
+    private static final Logger LOG =
+            LoggerFactory.getLogger(TaskExecutorStateChangelogStoragesManager.class);
+
+    /**
+     * This map holds all state changelog storages for tasks running on the task manager / executor
+     * that own the instance of this. Maps from job id to all the subtask's state changelog
+     * storages.
+     */
+    private final Map<JobID, StateChangelogStorage<?>> changelogStoragesByJobId;
+
+    private boolean closed;
+
+    /** shutdown hook for this manager. */
+    private final Thread shutdownHook;
+
+    public TaskExecutorStateChangelogStoragesManager() {
+        this.changelogStoragesByJobId = new HashMap<>();
+        this.closed = false;
+
+        // register a shutdown hook
+        this.shutdownHook =
+                ShutdownHookUtil.addShutdownHook(this::shutdown, getClass().getSimpleName(), LOG);
+    }
+
+    public StateChangelogStorage<?> stateChangelogStorageForJob(
+            @Nonnull JobID jobId, Configuration configuration) {
+        if (closed) {
+            throw new IllegalStateException(
+                    "TaskExecutorStateChangelogStoragesManager is already closed and cannot "
+                            + "register a new StateChangelogStorage.");
+        }
+
+        StateChangelogStorage<?> stateChangelogStorage = changelogStoragesByJobId.get(jobId);
+
+        if (stateChangelogStorage == null) {
+            stateChangelogStorage = StateChangelogStorageLoader.load(configuration);
+
+            changelogStoragesByJobId.put(jobId, stateChangelogStorage);
+
+            if (stateChangelogStorage != null) {
+                LOG.debug(
+                        "Registered new state changelog storage for job {} : {}.",
+                        jobId,
+                        stateChangelogStorage);
+            } else {
+                LOG.info(
+                        "Try to registered new state changelog storage for job {},"
+                                + "but result is null.",
+                        jobId);
```

Comment:
(please see the comment above about misconfiguration first)

I think if we allow nullable storage then we'll get a warning about it twice (here and in `StateChangelogStorageLoader`), plus an exception from backend or writer.

Reply:
I think it's OK for some redundant log when error occurs. 😄